File: baakjoon/2020/bj1965.c
#include <stdio.h>

int box[1001] = {0};
int dp[1001] = {0};
int box_dp(int n);

int main(){
    int i,n;
    scanf("%d",&n);
    for(i=0;i<n;i++){
        scanf("%d",&box[i]);
    }
    printf("%d",box_dp(n));
}

int box_dp(int n){
    int i,j,max;
    max = 1;
    for(i=0;i<n;i++){
        dp[i] = 1;
        for(j=0;j<i;j++){
            if(dp[i] < dp[j]+1 && box[j] < box[i]){
                dp[i] = dp[j]+1;
                if(max < dp[i]){
                    max = dp[i];
                }
            }
        }
    }
    return max;
}

File: baakjoon/2020/bj1965.py
import sys
sys.setrecursionlimit(10000000) 

def max_box(box,answer,now,n):
    if(now == n-1):
        return 1
    now_max = 1
    if(answer[now] != -1):
        return answer[now]
    for i in range(now,n):
        if(box[now] < box[i]):
            tmp = 1 + max_box(box,answer,i,n)
            if(now_max < tmp):
                now_max = tmp
    answer[now] = now_max
    return now_max

n = int(input(""))
box = list(map(int, input().split()))
answer = [-1] * (n+1)

for i in range(0,n):
    max_box(box,answer,i,n)
print(max(answer))
